### ISSUE TITLE
Failed json parsing in threads of ansible-container

### ISSUE THREAD

**Issue by haad** [ansible/ansible-container#178]
##### ISSUE TYPE
 - Bug Report

##### container.yml 
version: "1"
services:
   horizon:
     image: ubuntu:trusty
     ports:
       - "50001:50001"
     command: ['/usr/bin/dumb-init', '/usr/sbin/apache2ctl', '-D', 'FOREGROUND']
     dev_overrides:
       environment:
         - "DEBUG=1"

registries: {}
```

```

##### main.yml 
- name: Install python if there is none on our container
  hosts: all
  gather_facts: false
  tasks:
    - raw: which python || apt-get update
    - raw: (which python && which aptitude) || apt-get install -y python python-apt aptitude```

```

##### OS / ENVIRONMENT
Ansible Container, version 0.1.0
Darwin, ceres.local, 15.6.0, Darwin Kernel Version 15.6.0: Thu Jun 23 18:25:34 PDT 2016; root:xnu-3248.60.10~1/RELEASE_X86_64, x86_64
2.7.12 (default, Aug 15 2016, 10:56:05)
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)] /usr/local/opt/pyenv/versions/2.7.12/bin/python2.7
No DOCKER_HOST environment variable found. Assuming UNIX socket at /var/run/docker.sock
{u'Architecture': u'x86_64',
 u'BridgeNfIp6tables': True,
 u'BridgeNfIptables': True,
 u'CPUSet': True,
 u'CPUShares': True,
 u'CgroupDriver': u'cgroupfs',
 u'ClusterAdvertise': u'',
 u'ClusterStore': u'',
 u'Containers': 6,
 u'ContainersPaused': 0,
 u'ContainersRunning': 0,
 u'ContainersStopped': 6,
 u'CpuCfsPeriod': True,
 u'CpuCfsQuota': True,
 u'Debug': True,
 u'DefaultRuntime': u'runc',
 u'DockerRootDir': u'/var/lib/docker',
 u'Driver': u'aufs',
 u'DriverStatus': [[u'Root Dir', u'/var/lib/docker/aufs'],
                   [u'Backing Filesystem', u'extfs'],
                   [u'Dirs', u'36'],
                   [u'Dirperm1 Supported', u'true']],
 u'ExecutionDriver': u'',
 u'ExperimentalBuild': False,
 u'HttpProxy': u'',
 u'HttpsProxy': u'',
 u'ID': u'ED6O:F6NT:I2ET:63DW:ZCNI:6JMM:PYOB:WMCW:G7XH:XBJ2:JNVX:QB6Z',
 u'IPv4Forwarding': True,
 u'Images': 8,
 u'IndexServerAddress': u'https://index.docker.io/v1/',
 u'KernelMemory': True,
 u'KernelVersion': u'4.4.15-moby',
 u'Labels': None,
 u'LiveRestoreEnabled': False,
 u'LoggingDriver': u'json-file',
 u'MemTotal': 2097557504,
 u'MemoryLimit': True,
 u'NCPU': 4,
 u'NEventsListener': 1,
 u'NFd': 17,
 u'NGoroutines': 29,
 u'Name': u'moby',
 u'NoProxy': u'*.local, 169.254/16',
 u'OSType': u'linux',
 u'OomKillDisable': True,
 u'OperatingSystem': u'Alpine Linux v3.4',
 u'Plugins': {u'Authorization': None,
              u'Network': [u'bridge', u'null', u'host', u'overlay'],
              u'Volume': [u'local']},
 u'RegistryConfig': {u'IndexConfigs': {u'docker.io': {u'Mirrors': None,
                                                      u'Name': u'docker.io',
                                                      u'Official': True,
                                                      u'Secure': True}},
                     u'InsecureRegistryCIDRs': [u'127.0.0.0/8'],
                     u'Mirrors': None},
 u'Runtimes': {u'runc': {u'path': u'docker-runc'}},
 u'SecurityOptions': [u'seccomp'],
 u'ServerVersion': u'1.12.0',
 u'SwapLimit': True,
 u'Swarm': {u'Cluster': {u'CreatedAt': u'0001-01-01T00:00:00Z',
                         u'ID': u'',
                         u'Spec': {u'CAConfig': {},
                                   u'Dispatcher': {},
                                   u'Orchestration': {},
                                   u'Raft': {},
                                   u'TaskDefaults': {}},
                         u'UpdatedAt': u'0001-01-01T00:00:00Z',
                         u'Version': {}},
            u'ControlAvailable': False,
            u'Error': u'',
            u'LocalNodeState': u'inactive',
            u'Managers': 0,
            u'NodeAddr': u'',
            u'NodeID': u'',
            u'Nodes': 0,
            u'RemoteManagers': None},
 u'SystemStatus': None,
 u'SystemTime': u'2016-08-17T12:15:55.190805648Z'}
{u'ApiVersion': u'1.24',
 u'Arch': u'amd64',
 u'BuildTime': u'2016-07-28T21:15:28.963402499+00:00',
 u'GitCommit': u'8eab29e',
 u'GoVersion': u'go1.6.3',
 u'KernelVersion': u'4.4.15-moby',
 u'Os': u'linux',
 u'Version': u'1.12.0'}
```

```
##### SUMMARY
After successfull ansible-container build errors below are printed.

##### STEPS TO REPRODUCE
Run ansible-container build


##### ACTUAL RESULTS

```
Stopping ansible_horizon_1 ... done
Exporting built containers as images...
Committing image...
#!@# {"status":"tag","id":"sha256:f17e53c28cd1afdfb43f27b0a566f3746d5dc440f07744b8da17bdc7f2faa1b3","Type":"image","Action":"tag","Actor":{"ID":"sha256:f17e53c28cd1afdfb43f27b0a566f3746d5dc440f07744b8da17bdc7f2faa1b3","Attributes":{"com.docker.compose.config-hash":"77ecd2593999c2faa773c196d76af23cea50851353a3486d8cc3e925861eb5b2","com.docker.compose.container-number":"1","com.docker.compose.oneoff":"False","com.docker.compose.project":"ansible","com.docker.compose.service":"horizon","com.docker.compose.version":"1.7.0","name":"docker-gen-horizon:20160817115402"}},"time":1471434842,"timeNano":1471434842909692237}
{"status":"commit","id":"105fea13d3e26e3ebced76e49d4f1603595c91f19ae94a08cad43eb06d83ed69","from":"docker-gen-horizon:latest","Type":"container","Action":"commit","Actor":{"ID":"105fea13d3e26e3ebced76e49d4f1603595c91f19ae94a08cad43eb06d83ed69","Attributes":{"com.docker.compose.config-hash":"77ecd2593999c2faa773c196d76af23cea50851353a3486d8cc3e925861eb5b2","com.docker.compose.container-number":"1","com.docker.compose.oneoff":"False","com.docker.compose.project":"ansible","com.docker.compose.service":"horizon","com.docker.compose.version":"1.7.0","comment":"Built using Ansible Container","image":"docker-gen-horizon:latest","name":"ansible_horizon_1"}},"time":1471434842,"timeNano":1471434842909851644}

#!@# {"status":"tag","id":"sha256:f17e53c28cd1afdfb43f27b0a566f3746d5dc440f07744b8da17bdc7f2faa1b3","Type":"image","Action":"tag","Actor":{"ID":"sha256:f17e53c28cd1afdfb43f27b0a566f3746d5dc440f07744b8da17bdc7f2faa1b3","Attributes":{"com.docker.compose.config-hash":"77ecd2593999c2faa773c196d76af23cea50851353a3486d8cc3e925861eb5b2","com.docker.compose.container-number":"1","com.docker.compose.oneoff":"False","com.docker.compose.project":"ansible","com.docker.compose.service":"horizon","com.docker.compose.version":"1.7.0","name":"docker-gen-horizon:20160817115402"}},"time":1471434842,"timeNano":1471434842909692237}
{"status":"commit","id":"105fea13d3e26e3ebced76e49d4f1603595c91f19ae94a08cad43eb06d83ed69","from":"docker-gen-horizon:latest","Type":"container","Action":"commit","Actor":{"ID":"105fea13d3e26e3ebced76e49d4f1603595c91f19ae94a08cad43eb06d83ed69","Attributes":{"com.docker.compose.config-hash":"77ecd2593999c2faa773c196d76af23cea50851353a3486d8cc3e925861eb5b2","com.docker.compose.container-number":"1","com.docker.compose.oneoff":"False","com.docker.compose.project":"ansible","com.docker.compose.service":"horizon","com.docker.compose.version":"1.7.0","comment":"Built using Ansible Container","image":"docker-gen-horizon:latest","name":"ansible_horizon_1"}},"time":1471434842,"timeNano":1471434842909851644}

Exception in thread Thread-9:
Traceback (most recent call last):
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/site-packages/compose/cli/log_printer.py", line 190, in watch_events
    for event in event_stream:
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/site-packages/compose/project.py", line 343, in events
    decode=True
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/site-packages/docker/client.py", line 234, in _stream_helper
    data = json.loads(data)
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/json/decoder.py", line 367, in decode
    raise ValueError(errmsg("Extra data", s, end, len(s)))
ValueError: Extra data: line 2 column 1 - line 3 column 1 (char 615 - 1323)
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/site-packages/compose/cli/log_printer.py", line 190, in watch_events
    for event in event_stream:
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/site-packages/compose/project.py", line 343, in events
    decode=True
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/site-packages/docker/client.py", line 234, in _stream_helper
    data = json.loads(data)
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/local/opt/pyenv/versions/2.7.12/lib/python2.7/json/decoder.py", line 367, in decode
    raise ValueError(errmsg("Extra data", s, end, len(s)))
ValueError: Extra data: line 2 column 1 - line 3 column 1 (char 615 - 1323)

Exported docker-gen-horizon with image ID sha256:f17e53c28cd1afdfb43f27b0a566f3746d5dc440f07744b8da17bdc7f2faa1b3
Cleaning up horizon build container...
Cleaning up Ansible Container builder...
```

Problem seems to be in that data string from contains multiple root elements.

```
{"status":"tag","id":"sha256:f17e53c28cd1afdfb43f27b0a566f3746d5dc440f07744b8da17bdc7f2faa1b3","Type":"image","Action":"tag","Actor":{"ID":"sha256:f17e53c28cd1afdfb43f27b0a566f3746d5dc440f07744b8da17bdc7f2faa1b3","Attributes":{"com.docker.compose.config-hash":"77ecd2593999c2faa773c196d76af23cea50851353a3486d8cc3e925861eb5b2","com.docker.compose.container-number":"1","com.docker.compose.oneoff":"False","com.docker.compose.project":"ansible","com.docker.compose.service":"horizon","com.docker.compose.version":"1.7.0","name":"docker-gen-horizon:20160817115402"}},"time":1471434842,"timeNano":1471434842909692237}

{"status":"commit","id":"105fea13d3e26e3ebced76e49d4f1603595c91f19ae94a08cad43eb06d83ed69","from":"docker-gen-horizon:latest","Type":"container","Action":"commit","Actor":{"ID":"105fea13d3e26e3ebced76e49d4f1603595c91f19ae94a08cad43eb06d83ed69","Attributes":{"com.docker.compose.config-hash":"77ecd2593999c2faa773c196d76af23cea50851353a3486d8cc3e925861eb5b2","com.docker.compose.container-number":"1","com.docker.compose.oneoff":"False","com.docker.compose.project":"ansible","com.docker.compose.service":"horizon","com.docker.compose.version":"1.7.0","comment":"Built using Ansible Container","image":"docker-gen-horizon:latest","name":"ansible_horizon_1"}},"time":1471434842,"timeNano":1471434842909851644}

```

**Comment by j00bar** [comment]
We're not presently supporting Docker 1.12

**Comment by chouseknecht** [comment]
@haad based on the information you provided, everything seems to work. I tested your *main.yml* and *container.yml* in my environment. What follows are the results of my test. 

Note that I'm running the latest Ansible Container from source and Docker Machine 1.11.2.

```
$ ac --debug version

$ ac --debug version
Ansible Container, version 0.2.0-pre
Darwin, Chriss-MacBook-Pro.local, 15.6.0, Darwin Kernel Version 15.6.0: Thu Jun 23 18:25:34 PDT 2016; root:xnu-3248.60.10~1/RELEASE_X86_64, x86_64
2.7.11 (default, Jan 22 2016, 08:29:18)
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)] /usr/local/opt/python/bin/python2.7
{u'Architecture': u'x86_64',
 u'BridgeNfIp6tables': True,
 u'BridgeNfIptables': True,
 u'CPUSet': True,
 u'CPUShares': True,
 u'CgroupDriver': u'cgroupfs',
 u'ClusterAdvertise': u'',
 u'ClusterStore': u'',
 u'Containers': 2,
 u'ContainersPaused': 0,
 u'ContainersRunning': 0,
 u'ContainersStopped': 2,
 u'CpuCfsPeriod': True,
 u'CpuCfsQuota': True,
 u'Debug': True,
 u'DockerRootDir': u'/mnt/sda1/var/lib/docker',
 u'Driver': u'aufs',
 u'DriverStatus': [[u'Root Dir', u'/mnt/sda1/var/lib/docker/aufs'],
                   [u'Backing Filesystem', u'extfs'],
                   [u'Dirs', u'60'],
                   [u'Dirperm1 Supported', u'true']],
 u'ExecutionDriver': u'',
 u'ExperimentalBuild': False,
 u'HttpProxy': u'',
 u'HttpsProxy': u'',
 u'ID': u'TNH2:JOIY:XYYD:WFYR:FTTG:Z4NR:QVRT:PRKH:N2Z6:3DH2:RKHV:3LTF',
 u'IPv4Forwarding': True,
 u'Images': 6,
 u'IndexServerAddress': u'https://index.docker.io/v1/',
 u'KernelMemory': True,
 u'KernelVersion': u'4.4.12-boot2docker',
 u'Labels': [u'provider=virtualbox'],
 u'LoggingDriver': u'json-file',
 u'MemTotal': 1044250624,
 u'MemoryLimit': True,
 u'NCPU': 1,
 u'NEventsListener': 0,
 u'NFd': 17,
 u'NGoroutines': 34,
 u'Name': u'default',
 u'NoProxy': u'',
 u'OSType': u'linux',
 u'OomKillDisable': True,
 u'OperatingSystem': u'Boot2Docker 1.11.2 (TCL 7.1); HEAD : a6645c3 - Wed Jun  1 22:59:51 UTC 2016',
 u'Plugins': {u'Authorization': None,
              u'Network': [u'host', u'bridge', u'null'],
              u'Volume': [u'local']},
 u'RegistryConfig': {u'IndexConfigs': {u'docker.io': {u'Mirrors': None,
                                                      u'Name': u'docker.io',
                                                      u'Official': True,
                                                      u'Secure': True}},
                     u'InsecureRegistryCIDRs': [u'127.0.0.0/8'],
                     u'Mirrors': None},
 u'ServerVersion': u'1.11.2',
 u'SwapLimit': True,
 u'SystemStatus': None,
 u'SystemTime': u'2016-09-02T03:04:47.689699533Z'}
{u'ApiVersion': u'1.23',
 u'Arch': u'amd64',
 u'BuildTime': u'2016-06-01T21:20:08.558909126+00:00',
 u'GitCommit': u'b9f10c9',
 u'GoVersion': u'go1.5.4',
 u'KernelVersion': u'4.4.12-boot2docker',
 u'Os': u'linux',
 u'Version': u'1.11.2'}
```

```
$ cat ansible/container.yml

version: "1"
services:
  horizon:
    image: ubuntu:trusty
    ports:
      - "50001:50001"
    command: ['/usr/bin/dumb-init', '/usr/sbin/apache2ctl', '-D', 'FOREGROUND']
    dev_overrides:
      environment:
        - "DEBUG=1"

registries: {}
```

```
$ cat ansible/main.yml

- name: Install python if there is none on our container
  hosts: all
  gather_facts: false
  tasks:
    - raw: which python || apt-get update
    - raw: (which python && which aptitude) || apt-get install -y python python-apt aptitude
```

```
$ ac build

Starting Docker Compose engine to build your images...
Attaching to ansible_ansible-container_1
Cleaning up Ansible Container builder...
No image found for tag fix_178-horizon:latest, so building from scratch
trusty: Pulling from library/ubuntu
862a3e9af0ae: Pull complete
6498e51874bf: Pull complete
159ebdd1959b: Pull complete
0fdbedd3771a: Pull complete
7a1f7116d1e3: Pull complete
Digest: sha256:5b5d48912298181c3c80086e7d3982029b288678fccabf2265899199c24d7f89
Status: Downloaded newer image for ubuntu:trusty
Attaching to ansible_horizon_1, ansible_ansible-container_1
ansible-container_1  |
ansible-container_1  | PLAY [Install python if there is none on our container] ************************
ansible-container_1  |
ansible-container_1  | TASK [raw] *********************************************************************
ansible-container_1  | ok: [horizon]
ansible-container_1  |
ansible-container_1  | TASK [raw] *********************************************************************
ansible-container_1  | ok: [horizon]
ansible-container_1  |
ansible-container_1  | PLAY RECAP *********************************************************************
ansible-container_1  | horizon                    : ok=2    changed=0    unreachable=0    failed=0
ansible-container_1  |
ansible_ansible-container_1 exited with code 0
Aborting on container exit...
Stopping ansible_horizon_1 ... done
Exporting built containers as images...
Committing image...
Exported fix_178-horizon with image ID sha256:6eddfbd8b2d0ab1ba4ad7befc357fa2b4a0753f5f1251ca1ebb83c1a8714ef8a
Cleaning up horizon build container...
Cleaning up Ansible Container builder...
```

**Comment by chouseknecht** [comment]
Closing this, as I don't think we need to address. If you have additional information and questions, please feel free to re-open.

**Comment by roderickrandolph** [comment]
I've hit this error intermittently a few times with the latest 0.2.0-pre source. The issue appears to be related to a bug in the docker-py (docker/docker-py/issues/1059) library which affects docker-compose (docker/compose/issues/3821) and by extension ansible-container. The bug has been observed in both Docker 1.12 and 1.11. Looks like the fix has already been identified, merged, and pending the docker-py 1.10 release. Perhaps we can reopen this issue and close it once docker-py 1.10 is released and ansible-container implements the new version?
